In [1]:
import requests

data = requests.get("https://raw.githubusercontent.com/SkillfactoryDS/Datasets/master/war_peace_processed.txt").text

data = data.split('\n')
data.remove('')
data = data +['[new chapter]']

data[:10]

ConnectionError: HTTPSConnectionPool(host='raw.githubusercontent.com', port=443): Max retries exceeded with url: /SkillfactoryDS/Datasets/master/war_peace_processed.txt (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x13003d010>: Failed to resolve 'raw.githubusercontent.com' ([Errno 8] nodename nor servname provided, or not known)"))

In [ ]:
word_set = set(data)
len(word_set)
word_set.discard('[new chapter]')
print('Общее количество слов: {}'.format(len(data)))
print('Общее количество уникальных слов: {}'.format(len(word_set)))

Общее количество слов: 300080
Общее количество уникальных слов: 38210


{'мойке',
 'обтирая',
 'провожали',
 'знатного',
 'пробуя',
 'стоявшим',
 'известном',
 'складку',
 'похвалила',
 'серая',
 'ступая',
 'разоряются',
 'воинственный',
 'пронзали',
 'анисим',
 'лавочник',
 'выдумал',
 'геройской',
 'сынов',
 'нем',
 'народом',
 'неохотой',
 'грацией',
 'кучером',
 'подоконнике',
 'тачках',
 'влезал',
 'луже',
 'киевским',
 'смотровой',
 'клубе',
 'великолепное',
 'свершившийся',
 'назначены',
 'несимметрично',
 'надобно',
 'впервые',
 'свободные',
 'организме',
 'подводите',
 'распространявший',
 'бегала',
 'штральзунде',
 'наташи',
 'звучный',
 'сокольниках',
 'отыскивании',
 'подумавши',
 'характеру',
 'желчным',
 'советчиков',
 'ничтожной',
 'стремимся',
 'связываю',
 'происходившая',
 'совесть',
 'тоненькой',
 'торжественного',
 'отставали',
 'посторонились',
 'непредвиденно',
 'меньшая',
 'препровождением',
 'закричала',
 'независимости',
 'похоти',
 'приездом',
 'загибаясь',
 'проходивший',
 'лорда',
 'охотницкую',
 'ростову',
 'помогли',
 'приподн

In [ ]:
word_dict = dict()
count_chapter = 0
for word in data:
    if word == '[new chapter]':
        count_chapter += 1
        continue
    if word not in word_dict:
        word_dict[word] = 1
    else:
        word_dict[word] += 1
word_dict
print('Количество глав: {}'.format(count_chapter))

for i,key in enumerate(word_dict):
    if i == 10:
        break
    print(key, word_dict[key])

Количество глав: 171
1 10
в 6997
два 215
раза 35
короче 3
и 14592
пять 61
раз 296
интереснее 4
2 12


KeyError: 1

In [ ]:
chapter_data = []
chapter_words = []
for word in data:
    if word == '[new chapter]':
        chapter_data.append(chapter_words)
        chapter_words = []
    else:
        chapter_words.append(word)
print('Вложенный список содержит {} внутренних списка'.format(len(chapter_data)))
print(chapter_data[0][:100])
chapter_data[1]

Вложенный список содержит 171 внутренних списка
['1', 'в', 'два', 'раза', 'короче', 'и', 'в', 'пять', 'раз', 'интереснее', '2', 'почти', 'нет', 'философических', 'отступлений', '3', 'в', 'сто', 'раз', 'легче', 'читать', 'весь', 'французский', 'текст', 'заменен', 'русским', 'в', 'переводе', 'самого', 'толстого', '4', 'гораздо', 'больше', 'мира', 'и', 'меньше', 'войны', '5', 'хеппи-энд', 'эти', 'слова', 'я', 'поместил', 'семь', 'лет', 'назад', 'на', 'обложку', 'предыдущего', 'издания', 'указав', 'в', 'аннотации', 'первая', 'полная', 'редакция', 'великого', 'романа', 'созданная', 'к', 'концу', '1866', 'года', 'до', 'того', 'как', 'толстой', 'переделал', 'его', 'в', '1867--1869', 'годах', '--', 'и', 'что', 'я', 'использовал', 'такие-то', 'публикации', 'думая', 'что', 'все', 'всё', 'знают', 'я', 'не', 'объяснил', 'откуда', 'взялась', 'эта', 'первая', 'редакция', 'я', 'оказался', 'неправ', 'и', 'в', 'результате', 'оголтелые', 'и']


['автора',
 'я',
 'пишу',
 'до',
 'сих',
 'пор',
 'только',
 'о',
 'князьях',
 'графах',
 'министрах',
 'сенаторах',
 'и',
 'их',
 'детях',
 'и',
 'боюсь',
 'что',
 'и',
 'вперед',
 'не',
 'будет',
 'других',
 'лиц',
 'в',
 'моей',
 'истории',
 'может',
 'быть',
 'это',
 'нехорошо',
 'и',
 'не',
 'нравится',
 'публике',
 'может',
 'быть',
 'для',
 'нее',
 'интереснее',
 'и',
 'поучительнее',
 'история',
 'мужиков',
 'купцов',
 'семинаристов',
 'но',
 'со',
 'всем',
 'моим',
 'желанием',
 'иметь',
 'как',
 'можно',
 'больше',
 'читателей',
 'я',
 'не',
 'могу',
 'угодить',
 'такому',
 'вкусу',
 'по',
 'многим',
 'причинам',
 'во-первых',
 'потому',
 'что',
 'памятники',
 'истории',
 'того',
 'времени',
 'о',
 'котором',
 'я',
 'пишу',
 'остались',
 'только',
 'в',
 'переписке',
 'и',
 'записках',
 'людей',
 'высшего',
 'круга',
 'грамотных',
 'даже',
 'интересные',
 'и',
 'умные',
 'рассказы',
 'которые',
 'мне',
 'удалось',
 'слышать',
 'слышал',
 'я',
 'только',
 'от',
 'людей',
 'тог

In [ ]:
chapter_words_count = []
for chapter_word in chapter_data:
    temp = {}
    for word in chapter_word:
        if word not in temp:
            temp[word] = 1
        else:
            temp[word] +=1 
    chapter_words_count.append(temp)
chapter_words_count[15]

{'князь': 4,
 'василий': 4,
 'исполнил': 1,
 'обещание': 1,
 'данное': 1,
 'им': 1,
 'на': 17,
 'вечере': 2,
 'у': 11,
 'анны': 4,
 'павловны': 3,
 'пожилой': 1,
 'даме': 1,
 'просившей': 1,
 'его': 14,
 'о': 8,
 'своем': 1,
 'единственном': 1,
 'сыне': 2,
 'борисе': 1,
 'нем': 2,
 'было': 4,
 'доложено': 1,
 'государю': 2,
 'и': 65,
 'не': 17,
 'в': 33,
 'пример': 1,
 'другим': 1,
 'он': 24,
 'был': 6,
 'переведен': 1,
 'гвардии': 1,
 'семеновский': 1,
 'полк': 1,
 'прапорщиком': 1,
 'но': 7,
 'адъютантом': 1,
 'или': 6,
 'состоящим': 1,
 'при': 2,
 'кутузове': 1,
 'борис': 1,
 'так': 14,
 'назначен': 1,
 'несмотря': 3,
 'все': 5,
 'хлопоты': 1,
 'происки': 1,
 'михайловны': 1,
 'вскоре': 1,
 'после': 1,
 'вечера': 1,
 'анна': 5,
 'михайловна': 5,
 'вернулась': 1,
 'москву': 3,
 'прямо': 1,
 'к': 9,
 'своим': 3,
 'богатым': 1,
 'родственникам': 1,
 'ростовым': 1,
 'которых': 3,
 'она': 9,
 'стояла': 1,
 'москве': 3,
 'с': 19,
 'детства': 1,
 'воспитывался': 1,
 'годами': 1,
 'живал': 

In [ ]:
for chapter_number, chapter_dict in enumerate(chapter_words_count):
    if chapter_number == 2:
        break
    print('-'*40)
    print('Chapter: {}'.format(chapter_number))
    print('-'*40)
    for j, word in enumerate(chapter_dict):
        if j == 10:
            break
        print(word,chapter_dict[word])
            

----------------------------------------
Chapter: 0
----------------------------------------
1 1
в 37
два 3
раза 1
короче 1
и 34
пять 1
раз 2
интереснее 1
2 1
----------------------------------------
Chapter: 1
----------------------------------------
автора 1
я 20
пишу 2
до 1
сих 1
пор 1
только 5
о 2
князьях 1
графах 1


Давайте резюмировать, что мы с вами уже получили:

word_set - множество из всех слов, которые есть в книге

count_chapter - количество глав в книге (171)

word_counts - словарь, ключами которого являются слова, а значениями - количество вхождений этих слов в книгу

chapter_data - список из 171 списка, где элементы вложенных списков - все слова из главы. Каждый список соответствует своей главе

chapter_words_count - список из 171 словаря, где ключи - слова, а значения - количество слов в главе. Каждый словарь соответствует своей главе

In [ ]:
target_word = 'гостья'
target_chapter = 15

our_tf = list()
for chapter_number, elemnt_chapter in enumerate(chapter_words_count):
    dict_tf = dict()
    for target_tf in elemnt_chapter:
        dict_tf[target_tf] = elemnt_chapter[target_tf] / len(chapter_data[chapter_number])
        
    our_tf.append(dict_tf)
our_tf
'человек' in chapter_words_count[3]

True

In [ ]:
count = 0
word = 'человек'
for chapter_number in range(len(chapter_data)):
    if word in chapter_data[chapter_number]:
        count +=1
count

115

In [ ]:
our_df = dict()
for word in word_set:
    count = 0
    for chapter_number in range(len(chapter_data)):
        if word in chapter_data[chapter_number]:
            count +=1
    our_df[word] = count / len(chapter_data)
our_df

{'мойке': 0.005847953216374269,
 'обтирая': 0.017543859649122806,
 'провожали': 0.005847953216374269,
 'знатного': 0.005847953216374269,
 'пробуя': 0.023391812865497075,
 'стоявшим': 0.011695906432748537,
 'известном': 0.023391812865497075,
 'складку': 0.011695906432748537,
 'похвалила': 0.005847953216374269,
 'серая': 0.005847953216374269,
 'ступая': 0.04678362573099415,
 'разоряются': 0.005847953216374269,
 'воинственный': 0.011695906432748537,
 'пронзали': 0.005847953216374269,
 'анисим': 0.005847953216374269,
 'лавочник': 0.005847953216374269,
 'выдумал': 0.011695906432748537,
 'геройской': 0.011695906432748537,
 'сынов': 0.011695906432748537,
 'нем': 0.43859649122807015,
 'народом': 0.011695906432748537,
 'неохотой': 0.005847953216374269,
 'грацией': 0.005847953216374269,
 'кучером': 0.011695906432748537,
 'подоконнике': 0.005847953216374269,
 'тачках': 0.005847953216374269,
 'влезал': 0.005847953216374269,
 'луже': 0.005847953216374269,
 'киевским': 0.005847953216374269,
 'смотро

In [ ]:
our_df['человек']
our_tf['человек']

TypeError: list indices must be integers or slices, not str

In [ ]:
from math import log
tf_idf_out = dict()
for chapter_number in range(len(chapter_data)):
    tf_idf_chapter =  dict()
    for word in chapter_data[chapter_number]:
        if word not in tf_idf_chapter:
            tf_idf_chapter[word] = our_tf[chapter_number][word] * log(1/our_df[word])
    tf_idf_out[chapter_number] = tf_idf_chapter
tf_idf_out[1]

{'автора': 0.009868836000964798,
 'я': 0.004019614360423446,
 'пишу': 0.01130302871081167,
 'до': 0.00046799185518569766,
 'сих': 0.004025222876735579,
 'пор': 0.00221243667934431,
 'только': 0.0005783031815565906,
 'о': 0.00042696980848454664,
 'князьях': 0.009868836000964798,
 'графах': 0.009868836000964798,
 'министрах': 0.009868836000964798,
 'сенаторах': 0.009868836000964798,
 'и': 0.003369442235115895,
 'их': 0.0011930304608750557,
 'детях': 0.008538419147682753,
 'боюсь': 0.004321097502123792,
 'что': 0.0016044963024361403,
 'вперед': 0.002868385419693747,
 'не': 0.0014440466721925263,
 'будет': 0.0006797923487919653,
 'других': 0.0019165620539560984,
 'лиц': 0.003473049992950971,
 'в': 0.0008022481512180702,
 'моей': 0.0032776897351583754,
 'истории': 0.00787186603894182,
 'может': 0.003817832302197983,
 'быть': 0.0015343545103733687,
 'это': 0.0005517852863251146,
 'нехорошо': 0.004671042908638099,
 'нравится': 0.0058775854411186644,
 'публике': 0.009868836000964798,
 'для': 0

In [ ]:
target_word = 'анна'
target_chapter = 5
tf_idf_out[target_chapter][target_word]

0.011795917432668182

In [ ]:
target_chapter = 3
list_count = list()
list_word = list(tf_idf_out[target_chapter])

while len(list_count) < 3:
    count = 0
    word_checker = None
    for number, word in enumerate(list_word):
        if count < tf_idf_out[target_chapter][word]:
            count = tf_idf_out[target_chapter][word]
            word_checker = number
    list_count.append(list_word.pop(word_checker))
list_count
    
    
    

['павловна', 'анна', 'функе']

ыыдорогая мама прщлравляю днем рожденичя успехов во ввеа